src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -0,0 +1,75 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+import is.hail.expr.{BaseIR, FilterColsIR, FilterRowsIR, MatrixRead, TableFilter, TableRead}
+
+object Simplify {
+  def apply(ir: BaseIR): BaseIR = {
+    RewriteBottomUp(ir, matchErrorToNone {
+      // optimze IR
+
+      case Let(n1, v, Ref(n2, _)) if n1 == n2 => v
+
+      // If(NA, ...) handled by FoldConstants
+      case If(True(), x, _) => x
+      case If(False(), _, x) => x
+
+      case If(c, cnsq, altr) if cnsq == altr =>
+        If(IsNA(c), NA(cnsq.typ), cnsq)
+
+      case Cast(x, t) if x.typ == t => x
+
+      case ArrayLen(MakeArray(args, _)) => I32(args.length)
+
+      case ArrayRef(MakeArray(args, _), I32(i)) => args(i)
+
```

Comment:
want to add one for ArrayFilter with a True literal? This would let us remove FilterEntries(True) with just a couple other rules.

Reply:
I guess this is totally independent so you don't have to do it. Maybe we should keep a google doc with a list of rule ideas?

Reply:
Done.  And yes, let's start a dev post.